File: Data Structures & Algorithms/binary-search/submission-3.java
class Solution {
    public int search(int[] nums, int target) {
        int l = 0;
        int r = nums.length - 1;

        while (l <= r) {
            int mid = l + (r - l) / 2;
            int midVal = nums[mid];

            if (midVal < target) {
                l = mid + 1;
            } else if (midVal > target) {
                r = mid - 1;
            } else {
                return mid;
            }
        }

        return -1;
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        if sum(nums) % 2:
            return False
        
        dp = set()
        target = sum(nums) // 2

        def dfs(i, curr):
            if i >= len(nums):
                return False
            
            if curr == target:
                return True
            elif curr > target:
                return False
            
            res = False
            if (i, curr) not in dp:
                res |= dfs(i + 1, curr + nums[i])
                res |= dfs(i + 1, curr)
                dp.add((i, curr))
                return res
            
            return False
        
        return dfs(0, 0)